File: mandinka-flash-card-quiz.py
import json
import random
import string

# Basic English stopwords — you can customize this list
STOPWORDS = {"a", "an", "the", "and", "or", "but", "of", "to", "from", "is", "in", "on", "at", "by", "for", "with", "that", "this", "it", "as", "be"}

def tokenize_and_filter(text):
    # Lowercase, remove punctuation, split
    tokens = text.lower().translate(str.maketrans('', '', string.punctuation)).split()
    return [word for word in tokens if word not in STOPWORDS]

def load_cards(json_path):
    with open(json_path, 'r', encoding='utf-8') as f:
        return json.load(f)

def run_flashcard_quiz(json_path, direction="mandinka->english", num_cards=10):
    data = load_cards(json_path)
    if not data or num_cards < 1:
        print("No cards available or invalid number of cards.")
        return

    cards = random.sample(data, min(num_cards, len(data)))
    correct = 0
    total = 0

    for i, entry in enumerate(cards, 1):
        mandinka = entry["mandinka_word"]
        english_list = [t["english_word"] for t in entry["english_translations"]]
        english = ', '.join(english_list)

        if direction == "mandinka->english":
            print(f"{i}. Mandinka: {mandinka}")
            guess = input("English: ").strip().lower()
            if guess == "quit":
                break

            user_words = tokenize_and_filter(guess)
            translation_word_lists = [tokenize_and_filter(e) for e in english_list]

            match_found = any(
                any(user_word in translation_words for user_word in user_words)
                for translation_words in translation_word_lists
            )

            if match_found:
                print("Correct!\n")
                correct += 1
            else:
                print(f"Answer: {english}\n")

        elif direction == "english->mandinka":
            print(f"{i}. English: {english}")
            guess = input("Mandinka: ").strip().lower()
            if guess == "quit":
                break

            user_words = tokenize_and_filter(guess)
            mandinka_words = tokenize_and_filter(mandinka)

            match_found = any(user_word in mandinka_words for user_word in user_words)

            if match_found:
                print("Correct!\n")
                correct += 1
            else:
                print(f"Answer: {mandinka}\n")

        else:
            print("Invalid direction. Use 'mandinka->english' or 'english->mandinka'.")
            return

        total = i
        print(f"Your score so far: {correct}/{total}")

    print(f"Final score: {correct}/{total if total else num_cards}")

# Example
if __name__ == "__main__":
    run_flashcard_quiz("dictionary_json/dictionary_mandinka.json", direction="mandinka->english", num_cards=5)

File: games_html/assets/js/flashcards/mandinka_flashcard_quiz.js

import fs from 'fs';
import readline from 'readline';

// English stopwords
const STOPWORDS = new Set([
  "a", "an", "the", "and", "or", "but", "of", "to", "from",
  "is", "in", "on", "at", "by", "for", "with", "that", "this",
  "it", "as", "be", "not", "were", "where", "which", "these", "there", "their", "here"
]);

// (Placeholder for Mandinka stopwords — to be customized)
const MANDINKA_STOPWORDS = new Set();

function tokenizeAndFilter(text, lang = "en") {
  const STOP = lang === "en" ? STOPWORDS : MANDINKA_STOPWORDS;
  const tokens = text.toLowerCase().replace(/[.,\/#!$%\^&\*;:{}=\-_`~()]/g, "").split(/\s+/);
  return tokens.filter(word => word && !STOP.has(word));
}

function loadCards(jsonPath) {
  try {
    const data = fs.readFileSync(jsonPath, 'utf-8');
    return JSON.parse(data);
  } catch (error) {
    console.error(`Error loading or parsing cards from ${jsonPath}:`, error.message);
    return null;
  }
}

async function ask(question, rl) {
  return new Promise(resolve => rl.question(question, answer => resolve(answer.trim().toLowerCase())));
}

async function runFlashcardQuiz(jsonPath, direction = "mandinka->english", numCards = 10) {
  const data = loadCards(jsonPath);
  if (!data || numCards < 1) {
    console.log("No cards available or invalid number of cards.");
    return;
  }

  console.log("📘 Starting flashcard quiz. Type 'quit' at any time to exit.\n");

  const cards = data.sort(() => 0.5 - Math.random()).slice(0, Math.min(numCards, data.length));
  let correct = 0;

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  for (let i = 0; i < cards.length; i++) {
    const entry = cards[i];
    const mandinka = entry.mandinka_word;
    const englishList = entry.english_translations.map(t => t.english_word);
    const english = englishList.join(', ');

    let guess;
    if (direction === "mandinka->english") {
      console.log(`\n${i + 1}. Mandinka: ${mandinka}`);
      guess = await ask("English: ", rl);
      if (guess === "quit") break;

      const userWords = tokenizeAndFilter(guess, "en");
      if (userWords.length === 0) {
        console.log("❌ Too vague or empty response. Try to be more specific.\n");
        continue;
      }

      const translationWordLists = englishList.map(e => tokenizeAndFilter(e, "en"));
      const matchFound = translationWordLists.some(translationWords =>
        userWords.some(userWord => translationWords.includes(userWord))
      );

      if (matchFound) {
        console.log("✅ Correct!\n");
        correct++;
      } else {
        console.log(`❌ Answer: ${english}\n`);
      }

    } else if (direction === "english->mandinka") {
      console.log(`\n${i + 1}. English: ${english}`);
      guess = await ask("Mandinka: ", rl);
      if (guess === "quit") break;

      const userWords = tokenizeAndFilter(guess, "en");
      if (userWords.length === 0) {
        console.log("❌ Too vague or empty response. Try to be more specific.\n");
        continue;
      }

      const mandinkaWords = tokenizeAndFilter(mandinka, "en");
      const matchFound = userWords.some(userWord => mandinkaWords.includes(userWord));

      if (matchFound) {
        console.log("✅ Correct!\n");
        correct++;
      } else {
        console.log(`❌ Answer: ${mandinka}\n`);
      }

    } else {
      console.log("Invalid direction. Use 'mandinka->english' or 'english->mandinka'.");
      break;
    }

    console.log(`Score so far: ${correct}/${i + 1}`);
  }

  rl.close();
  console.log(`\n🎓 Final score: ${correct}/${cards.length}`);
}

// Accept command-line args: path, direction, count
const [,, jsonPathArg, directionArg, countArg] = process.argv;
const jsonPath = jsonPathArg || "dictionary_json/dictionary_mandinka.json";
const direction = directionArg || "mandinka->english";
const numCards = parseInt(countArg) || 5;

runFlashcardQuiz(jsonPath, direction, numCards);
